packages/alfa-math/src/matrix.ts

```diff
@@ -1,63 +1,175 @@
+      r[i] = 1;
+      return r;
+    });
+  }
+
+  export function size(m: Matrix): [number, number] {
+    return [rows(m), columns(m)];
+  }
+
+  export function rows(m: Matrix): number {
+    return m.length;
+  }
+
+  export function columns(m: Matrix): number {
+    return m.length === 0 ? 0 : m[0].length;
+  }
+
+  export function row(m: Matrix, i: number): Array<number> {
+    return m[i];
+  }
+
+  export function column(m: Matrix, i: number): Array<number> {
+    return m.map((r) => r[i]);
+  }
+
+  export function add(m: Matrix, n: Matrix): Matrix {
+    return m.map((r, i) => r.map((v, j) => v + n?.[i]?.[j]));
+  }
+
+  export function subtract(m: Matrix, n: Matrix): Matrix {
+    return m.map((r, i) => r.map((v, j) => v - n?.[i]?.[j]));
+  }
+
+  /**
+   * @see https://en.wikipedia.org/wiki/Matrix_multiplication
+   */
+  export function multiply(m: Matrix, n: number | Matrix): Matrix {
+    return typeof n === "number"
+      ? m.map((r) => r.map((v) => v * n))
+      : m.map((r, i) =>
+          n[0].map((_, j) =>
+            r.reduce((s, _, k) => s + m[i][k] * n?.[k]?.[j], 0)
+          )
+        );
+  }
+
+  /**
+   * Compute the transpose of a matrix.
+   *
+   * @see https://en.wikipedia.org/wiki/Transpose
+   */
+  export function transpose(m: Matrix): Matrix {
+    return m.length === 0 ? m : m[0].map((_, i) => m.map((row) => row[i]));
+  }
 
-  export function multiply<
-    N extends number,
-    M extends number,
-    P extends number
-  >(a: Matrix<N, M>, b: Matrix<M, P>): Matrix<N, P>;
+  /**
+   * Compute the determinant of a non-empty, square matrix.
+   *
+   * @see https://en.wikipedia.org/wiki/Determinant
+   * @see https://en.wikipedia.org/wiki/Laplace_expansion
+   *
+   * @remarks
+   * This function uses Laplace expansion for computing the determinant which
+   * has a time complexity of O(n!) and is therefore not practical for large
+   * matrices.
+   */
+  export function determinant(m: Matrix): number {
+    switch (m.length) {
+      // The determinant of a 1x1 matrix is the single value of the matrix.
+      case 1:
+        return m[0][0];
+
+      // The determinant of a 2x2 matrix is fairly short to write out and so
+      // this is done as an optimization.
+      case 2:
+        return m[0]?.[0] * m[1]?.[1] - m[0]?.[1] * m[1]?.[0];
+
+      default:
+        return m.reduce((d, _, i) => d + m[0]?.[i] * cofactor(m, 0, i), 0);
+    }
+  }
 
-  export function multiply<
-    N extends number,
-    M extends number,
-    P extends number
-  >(a: Matrix<N, M>, b: Matrix<M, P> | Vector<M>): Matrix<N, P> | Vector<M> {
-    return math.multiply(a as math.MathType, b) as Matrix<N, P> | Vector<M>;
+  /**
+   * Compute the inverse of an invertible matrix.
+   *
+   * @see https://en.wikipedia.org/wiki/Invertible_matrix
+   * @see https://en.wikipedia.org/wiki/Cramers_rule
+   *
+   * @remarks
+   * This function uses Cramer's rule for computing the determinant which
```

Comment:
Are we using Cramer's rule or Laplace's expansion for determinant 😕 ?
Or is Cramer's rule for another operation?

Reply:
Bah, it should say "inverse", not "determinant" 🤦